scripts/format_code.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+A tool for manage source file formatting.
+
+Assembles a list of files via paths, directories, and/or git diffs and runs
+clang-format on them.  By default the tool will format the files in place.
+-noformat_file will run a check-only mode.
+
+Returns:
+   0: Either no files were found which need formatting, or if format_file was
+     enablled and the files were succesfully formatted.
+   1: Files which need formatting were found but weren't formatted
+     because -noformat_file was set.
+   2: The application wasn't configured properly and could not execute.
+"""
+import io
+import os
+import subprocess
+import sys
+
+from os import listdir
+from os.path import isfile, join
+from absl import app
+from absl import flags
+
+
+# Flag Definitions:
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_boolean("git_diff", False, "Use git-diff to assemble a file list")
+flags.DEFINE_string("git_range", "main..", "the range string when using "
+  "git-diff")
+flags.DEFINE_multi_string("file", None, "Append the filename to the list of "
+  "files to check.")
+flags.DEFINE_multi_string("d", None, 
+  "Append the directory to the file list to format.")
+flags.DEFINE_multi_string("dr", None, 
+  "Append the directory, and recurse through its children, and add their "
+  "contents to the list of files to check.")
+flags.DEFINE_boolean("format_file", True, "Format files in place")
+flags.DEFINE_boolean("verbose", False, "Execute in verbose mode")
+
+# Constants:
+
+# The list of file types to run clang-format on.   Used to filter out
+# results when searching across directories or git diffs.
+FILE_TYPE_EXTENSIONS = (".ccp", ".cc", ".c", ".h")
+
+# Common execution arguments across all clang-format executions.
+CLANG_FORMAT_BASE = "clang-format -style=file "
+
+# Flag which informs clang-format to alter the file in place, overwriting
+# the original file with thew newly formatted version.
+CLANG_FORMAT_PARM_IN_PLACE = "-i "
+
+# Flag which informs clang-format to use stdout to log the changes it would
+# make to a file.  Logging in XML format allows us to easily parse the results 
+# to detect if the file requires formatting changes.
+CLANG_FORMAT_PARAM_LOG_REPLACEMENTS = "-output-replacements-xml "
```

Comment:
One nicer way than doing this (having concatenated strings with spaces included) is to just build all the parameters in a list in the first place. That way you also won't have to split them on " " which could cause an issue if filenames or paths have spaces in them.

Reply:
Done.